qiskit_nature/second_q/mappers/bosonic_linear_mapper.py

```diff
@@ -0,0 +1,115 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The Linear Mapper for Bosons."""
+
+from __future__ import annotations
+import operator
+
+from functools import reduce
+
+import numpy as np
+
+from qiskit.opflow import PauliSumOp
+from qiskit.quantum_info import Pauli, SparsePauliOp
+
+from qiskit_nature.second_q.operators import BosonicOp
+from .bosonic_mapper import BosonicMapper
+
+
+class BosonicLinearMapper(BosonicMapper):
+    """
+    The Linear boson-to-qubit mapping.
+    This mapper implements the formula in Section II.C of: https://arxiv.org/abs/2108.04258
+
+    b_k^\\dagger -> sum(sqrt(n_k + 1) \\sigma_(n_k)^+\\sigma_(n_k + 1)^-, from n_k = 0 to n_k^max + 1)
+    where n_k^max is the truncation order (defined by the user)
+    In the following, we explicit the operators \\sigma^+ and \\sigma^- with the Pauli matrices:
+    \\sigma_(n_k)^+ := S_j^+ = 0.5 * (X_j + i*Y_j) ; \\sigma_(n_k)^- := S_j^- = 0.5 * (X_j - i*Y_j)
+
+    The length of the qubit register is: BosonicOp.num_mode * (BosonicLinearMapper.truncation + 1)
+
+    Generates a 'linear_encoding' of the Bosonic operator b_k^\\dagger, b_k to qubit operators
+        (linear combinations of pauli strings).
+        In this 'linear_encoding' each bosonic mode is represented via n_k^max + 1 qubits, where n_k^max
+        is the truncation of the mode (meaning the number of states used in the expansion of the mode,
+        or equivalently the state at which the maximum excitation can take place).
+        The mode |k> is then mapped to the occupation number vector
+        |0_{nk_^max}, 0_{nk_^max - 1}, ..., 0_{n_k + 1}, 1_{n_k}, 0_{n_k - 1}, ..., 0_{0_k}>
+    """
+
+    def _map_single(
+        self, second_q_op: BosonicOp, *, register_length: int | None = None
+    ) -> SparsePauliOp | PauliSumOp:
+        # If register_length was passed, override the one present in the BosonicOp
```

Comment:
Suggestion to be discussed later (Note: add warning for no use of PauliSumOp)

```suggestion
    ) -> SparsePauliOp:
        """Maps a :class:`~qiskit_nature.second_q.operators.SparseLabelOp`
        to a``SparsePauliOp``.

        Args:
            second_q_op: the ``SparseLabelOp`` to be mapped.
            register_length: when provided, this will be used to overwrite the ``register_length``
                attribute of the operator being mapped. This is possible because the
                ``register_length`` is considered a lower bound in a ``SparseLabelOp``.

        Returns:
            The qubit operator corresponding to the problem-Hamiltonian in the qubit space.
        """
```